source/server/listener_manager_impl.cc

```diff
@@ -346,12 +347,14 @@ bool ListenerManagerImpl::addOrUpdateListener(const envoy::api::v2::Listener& co
     // In this case we can just replace inline.
     ASSERT(workers_started_);
     new_listener->debugLog("update warming listener");
-    new_listener->setSocket((*existing_warming_listener)->getSocket());
+    new_listener->setSocket((*existing_warming_listener)->getSocket(),
+                            (*existing_warming_listener)->bindToPort());
     *existing_warming_listener = std::move(new_listener);
   } else if (existing_active_listener != active_listeners_.end()) {
     // In this case we have no warming listener, so what we do depends on whether workers
     // have been started or not. Either way we get the socket from the existing listener.
-    new_listener->setSocket((*existing_active_listener)->getSocket());
+    new_listener->setSocket((*existing_active_listener)->getSocket(),
+                            (*existing_active_listener)->bindToPort());
```

Comment:
What happens if the listener configuration has changed, but in such a way that it wants to modify a listener option that must be set before binding? Should an error be raised in that case? Close/unbind the listener temporarily so the option can be changed?

Reply:
I personally think that pre-bind options shouldn't be allowed to change on a listener, because that would require the socket to be closed and reopened. But I don't see a way to guarantee options aren't changed with this design.

Reply:
Closing risks dropping incoming connections during a listener reconfigure. Maybe we could use SO_REUSEPORT option to allow multiple listening sockets on the same address, and create a new socket, apply pre-bind options, bind, listen, and then close the old socket?

Reply:
@jrajahalme I'm not clear whether there would be problems when those multiple sockets have different options, e.g. one socket has SOL_IP,IP_TRANSPARENT and another not?

Anyway, my vote would be to go with your design as a purely internal implementation, but not expose the `bound` flag directly into the LDS API. And any feature that sets pre-bind options should expose specific fields in LDS, and check that those fields are not modified when updating a listener. We already do that for bind addresses, and I proposed to do that for the `transparent` field in https://github.com/envoyproxy/envoy/pull/2719.
What do you think?

Reply:
@rlenglet what do you mean by:

> However, in any case, I think we shouldn't expose the bound flag as-is in LDS, since it would make it harder to control whether an option is modified when a listener is modified.

Do you mean in the API for filters `setOptions` call don't pass a bound flag? If we don't do that I'm not sure how this change could work as-is? Sorry I'm confused.